fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanChargeWritePlatformServiceImpl.java

```diff
@@ -1049,13 +1058,24 @@ private LoanOverdueDTO applyChargeToOverdueLoanInstallment(final Loan loan, fina
         Integer feeInterval = chargeDefinition.feeInterval();
         final ScheduledDateGenerator scheduledDateGenerator = new DefaultScheduledDateGenerator();
         Map<Integer, LocalDate> scheduleDates = new HashMap<>();
-        final Long penaltyWaitPeriodValue = this.configurationDomainService.retrievePenaltyWaitPeriod();
-        final Long penaltyPostingWaitPeriodValue = this.configurationDomainService.retrieveGraceOnPenaltyPostingPeriod();
+        Long penaltyWaitPeriodValue = this.configurationDomainService.retrievePenaltyWaitPeriod();
+        Long penaltyPostingWaitPeriodValue = this.configurationDomainService.retrieveGraceOnPenaltyPostingPeriod();
+
+        // we only want grace to be applied on the first installment to be overdue which coule be the first or any other
+        // so , we can just check if the loan has any penalty charged yet or not
+        boolean doesPenaltyExist = loan.hasPenaltiesInRepaymentSchedules();
+        if (doesPenaltyExist) {
```

Comment:
@oluexpert99 This check will never apply grace period in case when loan goes to arrears, comes out of arrears due to repayment and then goes back to arrears

Reply:
@faheem205  , maybe we can have a chat with Franco again on this . once penalties have been generated using the grace period once even after making payments , You can not enjoy the grace period again . 
I tested by creating a loan in arrears , then making payment of the first installment before running the penalty job . 
This allows the grace period to be applied on the second instalment only

Reply:
@oluexpert99 i still do not see how grace will be applied to an instalent when loan goes into arrears the second  time. 
Please consider the scenario. Loan is disbursed and first installment goes into arrears and enjoys grace period. 2nd installment also goes into arrears but no grace is applied. Now after payment, first installment is paid. Loan goes again to arrears and now ilgrace must be applied on installment 2

Reply:
Hi @faheem205  , you might need to check again . It was implemented in LoanRepaymentScheduleInstallment 

 public boolean hasPenalties() {
        return (this.penaltyCharges != null && this.penaltyCharges.compareTo(BigDecimal.ZERO) > 0) && !obligationsMet;
    }

Reply:
with this once an installement with grace is fully paid , a new installment can go into grace once in arrears